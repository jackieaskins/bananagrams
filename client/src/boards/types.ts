import { Tile } from '../tiles/types';

export enum Direction {
  ACROSS = 'ACROSS',
  DOWN = 'DOWN',
}
export type BoardPosition = {
  row: number;
  col: number;
};
export enum ValidationStatus {
  NOT_VALIDATED = 'NOT_VALIDATED',
  VALID = 'VALID',
  INVALID = 'INVALID',
}
export type WordInfo = {
  start: BoardPosition;
  validationStatus: ValidationStatus;
};
export type BoardSquare = {
  tile: Tile;
  validationStatus: ValidationStatus;
};
export type Board = Record<string, BoardSquare | null>;

export const getSquareId = ({ row, col }: BoardPosition): string =>
  `${row},${col}`;
