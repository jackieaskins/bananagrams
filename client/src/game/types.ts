import { Board } from '../board/types';
import { Hand } from '../hand/types';
import { Player } from '../player/types';
import { Tile } from '../tile/types';

export type GameStatus = 'NOT_STARTED' | 'STARTING' | 'IN_PROGRESS' | 'ENDING';
export type Snapshot = {
  players: Player[];
  hands: Record<string, Hand>;
  boards: Record<string, Board>;
};
export type GameInfo = {
  gameId: string;
  gameName: string;
  status: GameStatus;
  countdown: number;
  bunch: Tile[];
  players: Player[];
  hands: Record<string, Hand>;
  boards: Record<string, Board>;
  previousSnapshot: Snapshot | null;
};

export type GameLocationState = { gameInfo: GameInfo };
