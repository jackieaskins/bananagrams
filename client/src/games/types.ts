import { Player } from '../players/types';
import { Tile, TileItem } from '../tiles/types';
import { BoardLocation } from '../boards/types';

export type GameInfo = {
  gameId: string;
  gameName: string;
  isInProgress: boolean;
  bunch: Tile[];
  players: Player[];
  previousSnapshot: Player[] | null;
};

export type GameState = {
  gameInfo: GameInfo;
  handleDump: (tileItem: TileItem) => void;
  handleMoveTileFromBoardToHand: (boardLocation: BoardLocation | null) => void;
  handleMoveTileFromHandToBoard: (
    tileId: string,
    boardLocation: BoardLocation
  ) => void;
  handleMoveTileOnBoard: (
    fromLocation: BoardLocation,
    toLocation: BoardLocation
  ) => void;
  handlePeel: () => void;
  isInGame: boolean;
  walkthroughEnabled: boolean;
};

export type GameLocationState = Partial<GameState>;
