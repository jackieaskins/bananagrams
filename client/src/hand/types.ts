import { Tile } from '../tile/types';

export type Hand = Tile[];
