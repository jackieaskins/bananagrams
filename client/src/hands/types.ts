import { Tile } from '../tiles/types';

export type Hand = Record<string, Tile>;
