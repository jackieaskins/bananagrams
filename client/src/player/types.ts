export type Player = {
  userId: string;
  username: string;
  isReady: boolean;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
};
