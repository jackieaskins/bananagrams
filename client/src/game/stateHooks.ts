import { useRecoilCallback, useRecoilValue } from 'recoil';

import { Board } from '../boards/types';
import { initializeState } from '../game/state';
import { GameInfo, GameStatus } from '../games/types';
import { Hand } from '../hands/types';
import { Player } from '../players/types';

const {
  statusState,
  countdownState,
  nameState,
  playersState,
  currentPlayerState,
  handsState,
  currentHandState,
  boardsState,
  currentBoardState,
} = initializeState();

export const useGameStatus = (): GameStatus => useRecoilValue(statusState);

export const useGameCountdown = (): number => useRecoilValue(countdownState);

export const useGameName = (): string => useRecoilValue(nameState);

export const useGamePlayers = (): Player[] => useRecoilValue(playersState);
export const useCurrentPlayer = (): Player | null =>
  useRecoilValue(currentPlayerState);

export const useGameHands = (): Record<string, Hand> =>
  useRecoilValue(handsState);
export const useCurrentHand = (): Hand | null =>
  useRecoilValue(currentHandState);

export const useGameBoards = (): Record<string, Board> =>
  useRecoilValue(boardsState);
export const useCurrentBoard = (): Board | null =>
  useRecoilValue(currentBoardState);

export const useUpdateGameState = (): ((gameInfo: GameInfo) => void) =>
  useRecoilCallback(
    ({ set }) => (gameInfo) => {
      set(statusState, gameInfo.status);
      set(countdownState, gameInfo.countdown);
      set(nameState, gameInfo.gameName);
      set(playersState, gameInfo.players);
      set(handsState, gameInfo.hands);
      set(boardsState, gameInfo.boards);
    },
    []
  );
