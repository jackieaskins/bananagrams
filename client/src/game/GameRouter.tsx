import { useEffect } from 'react';

import { GameInfo } from '../games/types';
import { addGameInfoListener, removeGameInfoListener } from '../socket';
import Game from './Game';
import WaitingRoom from './WaitingRoom';
import { useGameStatus, useUpdateGameState } from './stateHooks';

type GameRouterProps = {
  initialGameInfo: GameInfo;
};

const GameRouter = ({ initialGameInfo }: GameRouterProps): JSX.Element => {
  const updateGameState = useUpdateGameState();
  const gameStatus = useGameStatus();
  const isGameInProgress = gameStatus === 'IN_PROGRESS';

  useEffect(() => {
    updateGameState(initialGameInfo);

    addGameInfoListener((info: GameInfo) => {
      updateGameState(info);
    });

    return (): void => {
      removeGameInfoListener();
    };
  }, [initialGameInfo, updateGameState]);

  if (isGameInProgress) {
    return <Game />;
  }

  return <WaitingRoom />;
};

export default GameRouter;
