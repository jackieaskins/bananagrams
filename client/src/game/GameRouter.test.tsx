import { shallow } from 'enzyme';
import { useEffect } from 'react';

import { GameInfo } from '../games/types';
import { addGameInfoListener, removeGameInfoListener } from '../socket';
import GameRouter from './GameRouter';
import { useIsGameInProgress } from './stateHooks';

jest.mock('react', () => ({
  ...jest.requireActual<any>('react'),
  useEffect: jest.fn().mockImplementation((f) => f()),
}));

jest.mock('../socket', () => ({
  addGameInfoListener: jest.fn(),
  removeGameInfoListener: jest.fn(),
}));

const mockUpdateGameState = jest.fn();
jest.mock('./stateHooks', () => ({
  useIsGameInProgress: jest.fn(),
  useUpdateGameState: () => mockUpdateGameState,
}));

const mockUseIsGameInProgress = useIsGameInProgress as jest.Mock;
const mockAddGameInfoListener = addGameInfoListener as jest.Mock;
const mockRemoveGameInfoListener = removeGameInfoListener as jest.Mock;
const mockUseEffect = useEffect as jest.Mock;
describe('<GameRouter />', () => {
  const initialGameInfo = ('initialGameInfo' as unknown) as GameInfo;

  const renderComponent = () =>
    shallow(<GameRouter initialGameInfo={initialGameInfo} />);

  beforeEach(() => {
    mockUseIsGameInProgress.mockReturnValue(false);
  });

  test('renders waiting room is game is not in progress', () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  test('renders properly if game is in progress', () => {
    mockUseIsGameInProgress.mockReturnValue(true);

    expect(renderComponent()).toMatchSnapshot();
  });

  describe('useEffect', () => {
    beforeEach(() => {
      renderComponent();
    });

    test('updates game state with initial info', () => {
      expect(mockUpdateGameState).toHaveBeenCalledWith(initialGameInfo);
    });

    test('adds game info listener', () => {
      const gameInfo = ('gameInfo' as unknown) as GameInfo;
      mockAddGameInfoListener.mock.calls[0][0](gameInfo);

      expect(mockAddGameInfoListener).toHaveBeenCalled();
      expect(mockUpdateGameState).toHaveBeenCalledWith(gameInfo);
    });

    test('removes game info listener on dismount', () => {
      mockUseEffect.mock.results[0].value();

      expect(mockRemoveGameInfoListener).toHaveBeenCalledWith();
    });
  });
});
