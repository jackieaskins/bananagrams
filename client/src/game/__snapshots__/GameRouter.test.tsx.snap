// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameRouter /> renders properly if game is in progress 1`] = `<Game />`;

exports[`<GameRouter /> renders waiting room is game is not in progress 1`] = `<WaitingRoom />`;
